modules/http/tfurlencode.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2018 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "plugin.h"
+#include "plugin-types.h"
+#include "template/simple-function.h"
+#include "messages.h"
+#include "cfg.h"
+
+#include "syslog-ng-config.h"
+
+#include <curl/curl.h>
+
+typedef struct
+{
+  TFSimpleFuncState super;
+  CURL *curl;
+} TFUrlEncodeState;
+
+gboolean
+tf_urlencode_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent,
+                     gint argc, gchar *argv[], GError **error)
+{
+
+  if (argc < 1)
+    {
+      msg_error("tfurlencode: missing mandatory parameter");
+      return FALSE;
+    }
+
+  TFUrlEncodeState *state = (TFUrlEncodeState *) s;
+  if (!(state->curl = curl_easy_init()))
+    {
+      msg_error("tfurlencode: cannot initialize libcurl");
+      return FALSE;
+    }
+
+  if (!tf_simple_func_prepare(self, state, parent, argc, argv, error))
+    {
+      g_set_error(error, LOG_TEMPLATE_ERROR, LOG_TEMPLATE_ERROR_COMPILE,
+                  "urlencode: prepare failed");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+tf_urlencode_call(LogTemplateFunction *self, gpointer s,
+                  const LogTemplateInvokeArgs *args, GString *result)
+{
+  TFUrlEncodeState *state = (TFUrlEncodeState *) s;
+  GString **argv = (GString **) args->bufs->pdata;
+  GString *input = argv[0];
+
+  gchar *urlencoded = curl_easy_escape(state->curl, input->str, input->len);
+  g_string_append(result, urlencoded);
+  g_free(urlencoded);
```

Comment:
As the libcurl documentation says it should be freed with `curl_free`, and I would recommend to capture it as a `char*` instead of `gchar*`.

Reply:
Nice catch thanks! I will fix it.